src/v/kafka/server/handlers/list_transactions.cc

```diff
@@ -0,0 +1,103 @@
+
+#include "kafka/server/handlers/list_transactions.h"
+
+#include "cluster/errc.h"
+#include "cluster/tm_stm.h"
+#include "cluster/tx_gateway_frontend.h"
+#include "cluster/types.h"
+#include "kafka/protocol/errors.h"
+#include "kafka/protocol/schemata/list_transactions_request.h"
+#include "kafka/server/handlers/details/security.h"
+#include "kafka/server/request_context.h"
+#include "kafka/server/response.h"
+#include "model/fundamental.h"
+#include "model/namespace.h"
+#include "model/record.h"
+#include "resource_mgmt/io_priority.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/when_all.hh>
+
+#include <algorithm>
+
+namespace kafka {
+
+namespace {
+
+bool filter_tx(
+  const list_transactions_request& req, const cluster::tm_transaction& tx) {
+    if (!req.data.producer_id_filters.empty()) {
+        if (std::none_of(
+              req.data.producer_id_filters.begin(),
+              req.data.producer_id_filters.end(),
+              [pid = tx.pid.get_id()](const auto& provided_pid) {
+                  return pid == provided_pid;
+              })) {
+            return false;
+        }
+    }
+
+    if (!req.data.producer_id_filters.empty()) {
+        if (std::none_of(
+              req.data.state_filters.begin(),
+              req.data.state_filters.end(),
+              [status = tx.get_status()](const auto& provided_status) {
+                  return status == provided_status;
+              })) {
+            return false;
+        }
+    }
+    return true;
+}
+
+} // namespace
+
+template<>
+ss::future<response_ptr>
+list_transactions_handler::handle(request_context ctx, ss::smp_service_group) {
+    list_transactions_request request;
+    request.decode(ctx.reader(), ctx.header().version);
+    vlog(klog.trace, "Handling request {}", request);
+
+    list_transactions_response response;
+
+    auto& tx_frontend = ctx.tx_gateway_frontend();
+    auto txs = co_await tx_frontend.get_all_transactions();
+    if (txs.has_value()) {
+        for (const auto& tx : txs.value()) {
+            if (!ctx.authorized(security::acl_operation::describe, tx.id)) {
+                // We should skip this transactional id
+                continue;
+            }
+
+            if (filter_tx(request, tx)) {
+                list_transaction_state tx_state;
+                tx_state.transactional_id = tx.id;
+                tx_state.producer_id = kafka::producer_id(tx.pid.id);
+                tx_state.transaction_state = ss::sstring(tx.get_status());
+                response.data.transaction_states.emplace_back(
+                  std::move(tx_state));
+            }
+        }
+    } else {
+        // In this 2 errors not coordinator got request and we just return empty
+        // array
+        if (
+          txs.error() != cluster::tx_errc::shard_not_found
+          && txs.error() != cluster::tx_errc::not_coordinator) {
+            response.data.error_code = kafka::error_code::unknown_server_error;
```

Comment:
we may want to avoid unknown_server_error,  Per KIP

```
COORDINATOR_LOAD_IN_PROGRESS: The coordinator is in the process of loading its state.
COORDINATOR_NOT_AVAILABLE: If the coordinator receiving the request is being shutdown.
```

Reply:
I am not sure we can understand us coordinator during shutdown or during loading_process/
This request is sent to each broker. If one of broker returns error client will fail request

Reply:
Kafka wraps some exception into unknown server error so even if it isn't mentioned in KIP it still may throw it. in this case it's a similar pathological situation